src/FindIntrinsics.cpp

```diff
@@ -894,14 +894,17 @@ Expr lower_halving_add(const Expr &a, const Expr &b) {
 
 Expr lower_halving_sub(const Expr &a, const Expr &b) {
     internal_assert(a.type() == b.type());
-    return (a >> 1) - (b >> 1) - (((b & 1) - (a & 1) + 1) >> 1);
+    Expr e = rounding_halving_add(a, ~b);
+    if (a.type().is_uint()) {
+        return e + make_const(e.type(), (uint64_t)1 << (a.type().bits() - 1));
```

Comment:
It's a bit hard to figure out why this works (took me a while), might be useful to have a comment to explain it (I got something like this, but maybe there is an easier way:

Assuming type is uint:

 ~b = 2 ** max_bits - b - 1

then 

rounding_halving_add(a, ~b) + 2 ** (max_bits -  1) = (a + (2 ** max_bits - b - 1) + 1) / 2 +  2 ** (max_bits -  1) = (a  - b + (2 ** max_bits)) / 2  +  2 ** (max_bits -  1) = (a - b) / 2  +  2 ** (max_bits -  1)   +  2 ** (max_bits -  1)  = (a - b) / 2  +  2 ** max_bits = (a - b) / 2  

).

Reply:
Good point. I added a comment.

How I actually discovered it was wondering if rounding_halving_add(x, ~y) was related to halving_sub(x, y), and noticing it was just off by 128 exactly.